File: apps/mobile/modules/encrypted-recording/ios/SegmentWriter.swift
import Foundation

struct SealedSegment {
  let index: Int
  let file: String
  let nonceBase64: String
  let sampleCount: Int64
  let durationMs: Int64
  let byteLength: Int64
  let tag: String
}

final class SegmentWriter {
  private let dir: URL
  private let crypto: SegmentCrypto
  private let sampleRate: Int
  private var index: Int
  private var handle: FileHandle?
  private var cryptor: SegmentCrypto.Cryptor?
  private var currentURL: URL?
  private var currentNonce: Data?
  private var sampleCount: Int64 = 0
  private var byteLength: Int64 = 0
  private var bytesSinceSync: Int64 = 0
  private var cipherAccum = Data()

  init(dir: URL, crypto: SegmentCrypto, sampleRate: Int, startIndex: Int = 0) {
    self.dir = dir
    self.crypto = crypto
    self.sampleRate = sampleRate
    self.index = startIndex - 1
  }

  var hasOpenSegment: Bool { handle != nil }
  var currentSampleCount: Int64 { sampleCount }
  var currentIndex: Int { index }

  private func randomName() -> String {
    var bytes = [UInt8](repeating: 0, count: 16)
    _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
    return bytes.map { String(format: "%02x", $0) }.joined()
  }

  private func intBE(_ value: Int) -> Data {
    let v = UInt32(truncatingIfNeeded: value)
    return Data([
      UInt8((v >> 24) & 0xFF), UInt8((v >> 16) & 0xFF),
      UInt8((v >> 8) & 0xFF), UInt8(v & 0xFF),
    ])
  }

  func openNext() {
    let nextIndex = index + 1
    let url = dir.appendingPathComponent(randomName())
    let nonce = crypto.newNonce()
    FileManager.default.createFile(atPath: url.path, contents: nil)
    guard let h = try? FileHandle(forWritingTo: url) else { return }
    h.write(nonce)
    let c = crypto.newCryptor(nonce: nonce)
    let encIndex = c.update(intBE(nextIndex))
    h.write(encIndex)
    cipherAccum = encIndex
    index = nextIndex
    handle = h
    cryptor = c
    currentURL = url
    currentNonce = nonce
    sampleCount = 0
    byteLength = 0
    bytesSinceSync = 0
  }

  func appendFrame(_ adts: Data, frameSamples: Int) {
    guard let h = handle, let c = cryptor else { return }
    let enc = c.update(adts)
    h.write(enc)
    cipherAccum.append(enc)
    sampleCount += Int64(frameSamples)
    byteLength += Int64(adts.count)
    bytesSinceSync += Int64(adts.count)
  }

  func maybeSync(threshold: Int64) {
    guard let h = handle else { return }
    if bytesSinceSync >= threshold {
      try? h.synchronize()
      bytesSinceSync = 0
    }
  }

  @discardableResult
  func seal() -> SealedSegment? {
    guard let h = handle else { return nil }
    let url = currentURL
    let nonce = currentNonce
    if let nonce = nonce {
      let tag = crypto.hmac(nonce: nonce, cipher: cipherAccum)
      h.write(tag)
    }
    try? h.synchronize()
    try? h.close()
    handle = nil
    cryptor = nil

    if sampleCount <= 0 || url == nil || nonce == nil {
      if let url = url { try? FileManager.default.removeItem(at: url) }
      currentURL = nil
      currentNonce = nil
      cipherAccum = Data()
      return nil
    }
    let tag = crypto.hmac(nonce: nonce!, cipher: cipherAccum)
    let sealed = SealedSegment(
      index: index,
      file: url!.lastPathComponent,
      nonceBase64: nonce!.base64EncodedString(),
      sampleCount: sampleCount,
      durationMs: sampleCount * 1000 / Int64(sampleRate),
      byteLength: byteLength,
      tag: tag.base64EncodedString()
    )
    currentURL = nil
    currentNonce = nil
    cipherAccum = Data()
    return sealed
  }
}

File: apps/mobile/modules/encrypted-recording/android/src/main/java/ai/multica/recording/crypto/SegmentWriter.kt
package ai.multica.recording.crypto

import java.io.File
import java.io.FileOutputStream
import java.security.SecureRandom
import javax.crypto.Cipher

data class SealedSegment(
  val index: Int,
  val file: String,
  val nonceBase64: String,
  val sampleCount: Long,
  val durationMs: Long,
  val byteLength: Long,
  val tag: String,
)

class SegmentWriter(
  private val dir: File,
  private val crypto: SegmentCrypto,
  private val sampleRate: Int,
  startIndex: Int = 0,
) {
  private val random = SecureRandom()
  private var index = startIndex - 1
  private var out: FileOutputStream? = null
  private var cipher: Cipher? = null
  private var currentFile: File? = null
  private var currentNonce: ByteArray? = null
  private var sampleCount = 0L
  private var byteLength = 0L
  private var bytesSinceSync = 0L
  private val cipherAccum = java.io.ByteArrayOutputStream()

  private fun randomName(): String {
    val bytes = ByteArray(16)
    random.nextBytes(bytes)
    return bytes.joinToString("") { "%02x".format(it) }
  }

  private fun intToBe(value: Int): ByteArray =
    byteArrayOf(
      (value ushr 24).toByte(),
      (value ushr 16).toByte(),
      (value ushr 8).toByte(),
      value.toByte(),
    )

  fun openNext() {
    val nextIndex = index + 1
    val file = File(dir, randomName())
    val nonce = crypto.newNonce()
    val stream = FileOutputStream(file)
    stream.write(nonce)
    val c = crypto.newCipher(nonce)
    val encIndex = c.update(intToBe(nextIndex))
    stream.write(encIndex)
    cipherAccum.reset()
    cipherAccum.write(encIndex)

    index = nextIndex
    out = stream
    cipher = c
    currentFile = file
    currentNonce = nonce
    sampleCount = 0
    byteLength = 0
    bytesSinceSync = 0
  }

  val hasOpenSegment: Boolean get() = out != null
  val currentIndex: Int get() = index
  val currentSampleCount: Long get() = sampleCount
  val currentFileName: String? get() = currentFile?.name

  fun appendFrame(adts: ByteArray, frameSamples: Int) {
    val stream = out ?: return
    val c = cipher ?: return
    val enc = c.update(adts)
    stream.write(enc)
    cipherAccum.write(enc)
    sampleCount += frameSamples
    byteLength += adts.size
    bytesSinceSync += adts.size
  }

  fun maybeSync(thresholdBytes: Long) {
    val stream = out ?: return
    if (bytesSinceSync >= thresholdBytes) {
      stream.flush()
      stream.fd.sync()
      bytesSinceSync = 0
    }
  }

  fun seal(): SealedSegment? {
    val stream = out ?: return null
    val c = cipher
    val file = currentFile
    val nonce = currentNonce
    try {
      if (c != null) {
        val tail = c.doFinal()
        if (tail.isNotEmpty()) {
          stream.write(tail)
          cipherAccum.write(tail)
        }
      }
      val tag = if (nonce != null) crypto.hmac(nonce, cipherAccum.toByteArray()) else ByteArray(0)
      if (tag.isNotEmpty()) stream.write(tag)
      stream.flush()
      stream.fd.sync()
    } finally {
      stream.close()
      out = null
      cipher = null
    }

    if (sampleCount <= 0L || file == null || nonce == null) {
      file?.delete()
      currentFile = null
      currentNonce = null
      cipherAccum.reset()
      return null
    }

    val result = SealedSegment(
      index = index,
      file = file.name,
      nonceBase64 = encodeB64(nonce),
      sampleCount = sampleCount,
      durationMs = sampleCount * 1000L / sampleRate,
      byteLength = byteLength,
      tag = encodeB64(crypto.hmac(nonce, cipherAccum.toByteArray().let { it })),
    )
    // tag already written; recompute from accum (includes tail)
    currentFile = null
    currentNonce = null
    cipherAccum.reset()
    return result
  }
}
